sysl2/sysl/integration/deps.go

```diff
@@ -0,0 +1,304 @@
+			outboundDeps.Deps[dep] = struct{}{}
+		}
+	}
+	if lenPassthroughs > 0 {
+		for dep := range outboundDeps.Deps {
+			walkPassthrough(excludes, passthroughs, dep, integrations, module)
+		}
+	}
+
+	return integrations
+}
+
+func (ds *DependencySet) ResolveDependencies(module *sysl.Module) {
+	apps := module.GetApps()
+	for appname, app := range apps {
+		for epname, endpoint := range app.GetEndpoints() {
+			ds.resolveStatementDependencies(endpoint.GetStmt(), apps, appname, epname)
+		}
+	}
+
+	if len(ds.errors) > 0 {
+		panic(fmt.Sprintf("broken deps:\n  %s", strings.Join(ds.errors, "\n")))
+	}
+}
+
+func (ds *DependencySet) resolveStatementDependencies(stmts []*sysl.Statement, apps map[string]*sysl.Application, appname, epname string) {
+	for _, stat := range stmts {
+		switch c := stat.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			var errStr string
+			targetName := getAppName(c.Call.GetTarget())
+			targetApp := apps[targetName]
+			if targetApp == nil {
+				errStr = fmt.Sprintf("%s <- %s: calls non-existent app %s", appname, epname, targetName)
+				ds.errors = append(ds.errors, errStr)
+			} else {
+				isValid := !hasPattern("abstract", utils.MakeStrSetFromSpecificAttr("patterns", targetApp.GetAttrs()).ToSlice())
+				if !isValid {
+					panic(fmt.Sprintf("call target '%s' must not be ~abstract", targetName))
+				}
+				callEndpoint := c.Call.GetEndpoint()
+				if targetApp.GetEndpoints()[callEndpoint] == nil {
+					errStr = fmt.Sprintf(
+						"%s <- %s: calls non-existent endpoint %s -> %s", appname, epname, targetName, callEndpoint)
+					ds.errors = append(ds.errors, errStr)
+				} else {
+					selfApp := MakeAppElement(appname, epname)
+					targetApp := MakeAppElement(targetName, callEndpoint)
+					dep := MakeAppDependency(selfApp, targetApp)
+					ds.Deps[dep] = struct{}{}
+				}
+			}
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			ds.resolveStatementDependencies(c.Cond.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Loop:
+			ds.resolveStatementDependencies(c.Loop.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_LoopN:
+			ds.resolveStatementDependencies(c.LoopN.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Foreach:
+			ds.resolveStatementDependencies(c.Foreach.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Group:
+			ds.resolveStatementDependencies(c.Group.GetStmt(), apps, appname, epname)
+		case *sysl.Statement_Alt:
+			for _, choice := range c.Alt.GetChoice() {
+				ds.resolveStatementDependencies(choice.GetStmt(), apps, appname, epname)
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func isSub(child, parent utils.StrSet) bool {
+	if len(child) == 0 && len(parent) == 0 {
+		return true
+	}
+	if len(parent) == 0 {
+		return false
+	}
+	if len(child) == 0 {
+		return true
+	}
+	return len(child.Difference(parent)) == 0
+}
+
+func hasPattern(s string, arr []string) bool {
```

Comment:
duplicate code. this function exists somewhere else too.

Reply:
I think it's not duplicate. I've checked that there is a function named as HasAbstractPattern() in helpers.go. It returns true if has the specific pattern for "abstract". However, what I want to is to check if it has any other pattern.

Reply:
@loki5265 then refactor that function.
the logic in your hasPattern is strange.
you are first looping through all elements to build a map and then using map lookup to find if the key exists.
you are looping anyway. 
so just loop to find what you are looking for.

Reply:
yep, refactored the original hasPattern function and removed this one.